cloud_storage: delete objects from S3 on deletion of tiered storage topics

jcsp opened the pull request:
## Cover letter

This behavior mirrors how we handle deletion of local data: each node that has a cluster::partition instance does a simple for loop through the manifest and deletes each segment.

This sends many redundant DELETE operations to S3, but is necessary for safety, as if we tried to e.g. only do this from the leader then we would be dependent on the leader being available at deletion time.

This behavior is _not_ applied to:
- Read replica topics
- Topics without both remote.read and remote.write set to true (i.e. legacy "archival" topics)

Still to add:
- [ ] DeleteObjects S3 API for efficient bulk deletion
- [x] Cover txrange objects

I think we will probably release 22.3 without plugging all the cases that can leave objects behind in S3: this will require implementing scrub of S3 data.

## Backport Required

- [x] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

None

## Release notes

### Features

* Topics using Tiered Storage will now clean up objects in S3 when they are deleted.  This may be avoided by disabling tiered storage on the topic before deleting it.

VladLazar commented on tests/rptest/tests/topic_delete_test.py:
```diff
@@ -73,6 +71,71 @@ def topic_storage_purged():
+        # Confirm objects in remote storage
+        before_objects = self.s3_client.list_objects(
+            self.si_settings.cloud_storage_bucket)
+        assert sum(1 for _ in before_objects) > 0
```
nit: `assert len(before_objects) > 0`?

VladLazar commented on tests/rptest/tests/read_replica_e2e_test.py:
```diff
@@ -175,18 +175,32 @@ def _bucket_delta(self, bu1: BucketUsage,
                 in str(e)):
             second_rpk.produce(self.topic_name, "", "test payload")
 
+        objects_before = sum(1 for _ in self.redpanda.s3_client.list_objects(
```
nit: Could you use `len` here?

jcsp replied:
list_objects is an iterator, so no len.  The `1 for _ in ` thing is a more efficient version for saying `len(list(the iterator))`

VladLazar replied:
Fair enough. I'll resolve those comments.

jcsp commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -800,4 +801,88 @@ remote_partition::iterator remote_partition::upper_bound(kafka::offset o) {
+    auto manifest_get_result = co_await _api.maybe_download_manifest(
+      _bucket, manifest_path, manifest, local_rtc);
+
+    if (manifest_get_result != download_result::notfound) {
```
That is a good point: this check should have been a three way think - throw on errors, skip deletion on notfound, and do deletion on success.

jcsp replied:
Fixed this and other error handling.

mmaslankaprv commented:
@jcsp are you going to add `redpanda.remote.topic-deletion-delay.hours` to it ?

jcsp commented:
> @jcsp are you going to add redpanda.remote.topic-deletion-delay.hours to it ?

The "recycle bin" feature is in the future -- for 22.3 there will be a boolean redpanda.remote.delete flag that defaults to false for legacy topics, and can be manually set to false for topics if user wants to delete a topic without destroying data in S3.

jcsp commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -800,4 +801,88 @@ remote_partition::iterator remote_partition::upper_bound(kafka::offset o) {
+        for (const auto& i : manifest) {
+            auto path = manifest.generate_segment_path(i.first, i.second);
+            vlog(_ctxlog.debug, "Erasing segment {}", path);
+            // On failure, we throw: this should cause controller to retry
+            // the topic deletion operation that called us, until it eventually
+            // succeeds.
```
I think the `retry_chain_node`'s timeout should implicitly limit the number of retries, although need to test the behavior (one of my test TODOs is to block access to S3 for a while while deleting a topic and check that's handled properly)

Lazin commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -800,4 +801,88 @@ remote_partition::iterator remote_partition::upper_bound(kafka::offset o) {
+            // longer exists, or the S3 backend is just permanently gone?
+            // TODO: S3 API has a plural delete API, which would be more
+            // suitable.
+            co_await _api.delete_object(
```
the return value of `delete_object` is ignored

jcsp replied:
Good spot, I've added error handling to these calls: treating timeouts as retryable  and other errors as a cue to give up (explaining why in comments)

Lazin commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -800,4 +801,88 @@ remote_partition::iterator remote_partition::upper_bound(kafka::offset o) {
+
+        // Erase the partition manifest
+        vlog(_ctxlog.debug, "Erasing partition manifest {}", manifest_path);
+        co_await _api.delete_object(
```
same here, it can timeout

jcsp commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -800,4 +801,88 @@ remote_partition::iterator remote_partition::upper_bound(kafka::offset o) {
+    // (TODO: fix this by implementing a scrub operation that finds objects
+    //  not mentioned in the manifest, and call this before erase())
+
+    static ss::lowres_clock::duration erase_timeout = 60s;
+    static ss::lowres_clock::duration erase_backoff = 1s;
```
They sure can!

jcsp commented on tests/rptest/tests/topic_delete_test.py:
```diff
@@ -73,6 +71,71 @@ def topic_storage_purged():
+        def complete():
+            after_objects = self.s3_client.list_objects(
+                self.si_settings.cloud_storage_bucket)
+            self.logger.debug("Objects after topic deletion:")
+            empty = True
+            for i in after_objects:
+                self.logger.debug(f"  {i}")
+                empty = False
+
+            return empty
```
This is another case where `after_objects` is an iterator rather than a list (i.e. wouldn't get a stringized list if passed to formatter)

jcsp commented on src/v/cloud_storage/remote.cc:
```diff
@@ -658,14 +658,13 @@ ss::future<download_result> remote::segment_exists(
 }
 
-ss::future<upload_result> remote::delete_segment(
+ss::future<upload_result> remote::delete_object(
```
Fixed.

jcsp commented:
Updated for comments thus far, and to add the new remote.delete topic property.

I've fleshed out error handling in a way that tries to avoid the controller getting stuck in nasty corner cases (like if someone deletes a bucket and then deletes topics, such that the topic delete can never successfully delete its segments).  The approach is to throw on timeouts (so that controller will retry applying the change), but log + proceed on other errors (on the assumption that they are permanent).

Some more work to do tomorrow:
- Fix an occasional failure to set remote.delete in the test (it's odd, the kafka request is handled properly, but somehow the configuration update comes through the controller backend with the wrong value.  I suspect maybe I'm hitting a race where ADL encoding is getting used if things happen too early)
- Consolidate all the places I default remote.delete to true in to some constant (I don't think we need to make it configurable but it shouldn't rely on all the places I default-initialize agreeing on the value).

I might push off doing DeleteObjects to a followup PR.

VladLazar commented on src/v/cluster/types.h:
```diff
@@ -1206,7 +1219,8 @@ struct incremental_topic_updates
           retention_local_target_bytes,
-          retention_local_target_ms);
+          retention_local_target_ms,
+          remote_delete);
```
You have to update the ADL encoding too in order not to lose updates.

`IncrementalAlterConfigs` requests are forwarded if they don't land on the controller leader:
https://github.com/redpanda-data/redpanda/blob/b5b4a022904a934836a1aae4976d12268bd49e80/src/v/cluster/topics_frontend.cc#L226-L230

If that's the first request on the connection it will be ADL encoded.

Q: ADL decoding for `incremental_topic_updates` is not forwards compatible? Why is it safe to evolve the encoding?
A: We refuse to update updating topic configuration while in a mixed version state. If the request was always processed on the leader we wouldn't need to update the ADL encoding with actual values.
https://github.com/redpanda-data/redpanda/blob/b5b4a022904a934836a1aae4976d12268bd49e80/src/v/cluster/topics_frontend.cc#L180-L190

jcsp replied:
Added this to the ADL encoding.

VladLazar reviewed (commented):
Two things and a few nits:
1. You have to update the ADL encoding for `incremental_topic_updates` to include the remote deletion field (see comment here)
2. I'm not sure about the way in which the default is applied. To me it looks like we default to `delete=true` for upgraded topics too. (see comment here)

jcsp commented on src/v/kafka/server/handlers/incremental_alter_configs.cc:
```diff
@@ -42,9 +42,9 @@ using resp_resource_t = incremental_alter_configs_resource_response;
  * resolved.
  */
-template<typename T>
+template<typename t>
```
accident, reverted.

jcsp commented on src/v/kafka/server/handlers/incremental_alter_configs.cc:
```diff
@@ -54,7 +54,36 @@ void parse_and_set_optional(
+
+    if (op == config_resource_operation::set && value) {
+        try {
+            property.value = string_switch<bool>(*(value))
```
fixed

jcsp commented on src/v/kafka/server/handlers/incremental_alter_configs.cc:
```diff
@@ -54,7 +54,36 @@ void parse_and_set_optional(
+            property.value = string_switch<bool>(*(value))
+                               .match("true", true)
+                               .match("false", false);
+        } catch (...) {
```
Yes, although it's still quite ugly (string_switch throws runtime_error rather than something nicely typed).  Still, you're right, it's not like I meant to catch e.g. bad_alloc here.  Fixed.

jcsp commented on src/v/storage/ntp_config.h:
```diff
@@ -165,6 +165,16 @@ class ntp_config {
+     */
+    bool is_tiered_storage() const {
+        return _overrides != nullptr && !_overrides->read_replica
+               && _overrides->shadow_indexing_mode
```
I think this code is still okay if `shadow_indexing_mode` is nullopt?  The equality operator returns false if you compare a value type to the optional

VladLazar replied:
You're right. It doesn't de-reference the optional. Resolving.

VladLazar commented on src/v/cluster/types.cc:
```diff
@@ -498,7 +500,7 @@ std::ostream& operator<<(std::ostream& o, const incremental_topic_updates& i) {
       "segment_size: {} retention_bytes: {} retention_duration: {} "
       "shadow_indexing: {}, batch_max_bytes: {}, retention_local_target_bytes: "
-      "{}, retention_local_target_ms: {}}}",
+      "{}, retention_local_target_ms: {}, remote.delete: {}}}",
```
nit: `remote_delete`?

jcsp replied:
fixed

VladLazar commented on src/v/storage/ntp_config.h:
```diff
@@ -165,6 +165,16 @@ class ntp_config {
+     * both reads and writes to S3, and is not a read replica.
+     */
+    bool is_tiered_storage() const {
+        return _overrides != nullptr && !_overrides->read_replica
```
I can't see the change. Did you miss it while rebasing?

VladLazar commented on tests/rptest/tests/topic_delete_test.py:
```diff
@@ -73,6 +73,94 @@ def topic_storage_purged():
+    @parametrize(disable_delete=False)
+    @parametrize(disable_delete=True)
+    def topic_delete_cloud_storage_test(self, disable_delete):
+        # Set retention to 5MB
+        self.kafka_tools.alter_topic_config(
+            self.topic, {'retention.bytes': 5 * 1024 * 1024})
```
`SISettings` enables remote write at the cluster level by default. To adjust the local retention you have to set `redpanda.remote.local.target.bytes`.

jcsp commented on src/v/cluster/partition.h:
```diff
@@ -283,6 +283,7 @@ class partition {
     ss::shared_ptr<cloud_storage::remote_partition> _cloud_storage_partition;
     ss::lw_shared_ptr<const storage::offset_translator_state> _translator;
     std::optional<s3::bucket_name> _read_replica_bucket{std::nullopt};
+    bool _remote_delete_enabled{true};
```
The behavior here is that the struct itself defaults to true, but the false value is applied when decoding from legacy versions.

VladLazar replied:
I understand, but I'm not sure the default is propagated for topic creation. `_remote_delete_enabled` is only updated via
`partition::update_configuration`, but that's not called on the topic creation path. I think this means that if a topic was created by a previous redpanda version and it's config was never updated, deletion would be enabled.

jcsp replied:
The creation case is in the `topic_properties` adl decoder (create topics is a list of `topic_configuration`, each of which as a `topic_properties`).  That's covered in `test_retention_config_on_upgrade_from_v22_2_to_v22_3` (checks that remote.delete is false on topics created on old version.

VladLazar replied:
`test_retention_config_on_upgrade_from_v22_2_to_v22_3` essentially tests that the `remote_delete` property in the topic table has the expected value. The bit I don't see is how `_remote_delete_enabled` gets initialised on creation.

jcsp commented on src/v/cluster/partition.cc:
```diff
@@ -375,7 +375,11 @@ partition::timequery(storage::timequery_config cfg) {
-    return _raft->log().update_configuration(
+    // XXX remove
+    vlog(clusterlog.trace, "update_configuration {}", properties);
+    _remote_delete_enabled = properties.remote_delete;
```
Let's punt on this until we revisit ntp_config vs. cloud storage stuff

jcsp commented:
Force push:
- Added remote.delete to adl encoding
- Added coverage of the property's behavior in the topic_creation_test upgrade path
- Various nits

In the process of testing this I touched on incremental config behavior and found we have a more general issue (https://github.com/redpanda-data/redpanda/issues/6773), which we'll need to address separately before GA.

jcsp commented:
Force push for some nits in the compat testing: it needs overriding for cases where it expects to be able to round-trip values through ADL, but we sometimes say "ADL decode == remote delete is false", and for updating the TopicSpec definition to know about the new property (apparently there are tests that take all topic properties and try to shove them into the TopicSpec constructor)

jcsp commented:
This is so weird. The serialization roundtrip parts of test_cluster_rpunit are failing _only_ in release mode, and did so somewhat reliably (both in previous CI run and this one, after I'd made a change to avoid injecting random incremental updates for remote_delete), but when I run locally 100x repeats of the tests on a release build, I get no failures.

jcsp commented:
/ci-repeat 5

jcsp commented:
Repeats:
- 1 instance of https://github.com/redpanda-data/redpanda/issues/6792
- 2 failures of ShadowIndexingCacheSpaceLeakTest with "file not found" BadLogLines.  This is https://github.com/redpanda-data/redpanda/issues/6601, but that's a suspiciously frequent reproduction so I'll investigate.
- 1 failure of `test_writes_forbidden` that I think is an overly strict assertion criterion, will check + fix.
- 1 failure of EndToEndTopicRecovery.test_restore with a consumer timeout, will investigate.

mmaslankaprv commented on src/v/kafka/server/handlers/create_topics.cc:
```diff
@@ -45,6 +45,7 @@ static constexpr std::array<std::string_view, 14> supported_configs{
   topic_property_recovery,
   topic_property_remote_write,
   topic_property_remote_read,
+  topic_property_remote_delete,
```
also this feature is missing a test case

VladLazar commented:
Looks good to me. My only outstanding comment is this: https://github.com/redpanda-data/redpanda/pull/6683#discussion_r994401385. Maybe, I'm missing something and it actually works fine, but a simple ducktape test that creates a topic with `remote.delete=false` and checks that no data is removed from S3, would still be nice.

VladLazar reviewed (commented):
Looks good to me. My only outstanding comment is this: https://github.com/redpanda-data/redpanda/pull/6683#discussion_r994401385. Maybe, I'm missing something and it actually works fine, but a simple ducktape test that creates a topic with remote.delete=false and checks that no data is removed from S3, would still be nice.

mmaslankaprv commented on src/v/kafka/server/handlers/alter_configs.cc:
```diff
@@ -47,6 +47,24 @@ void parse_and_set_optional(
+        property_update.value = default_value;
+    } else {
+        try {
+            property_update.value = string_switch<bool>(*(value))
```
nit: can we just do `boost::lexical_cast<bool>(*(value))` ?

mmaslankaprv commented on src/v/cluster/partition.cc:
```diff
@@ -414,6 +416,28 @@ ss::future<> partition::remove_persistent_state() {
     if (_id_allocator_stm) {
         co_await _id_allocator_stm->remove_persistent_state();
     }
+
+    // Backward compatibility: even if remote.delete is true, only do
+    // deletion if the partition is in full tiered storage mode (this
+    // excludes read replica clusters from deleting data in S3)
+    bool tiered_storage
+      = get_ntp_config().is_tiered_storage()
+        || (config::shard_local_cfg().cloud_storage_enable_remote_write() &&
+            config::shard_local_cfg().cloud_storage_enable_remote_read());
+
+    if (_cloud_storage_partition && tiered_storage && _remote_delete_enabled) {
+        vlog(
+          clusterlog.debug,
+          "Erasing S3 objects for partition {} ({} {} {})",
+          ntp(),
+          get_ntp_config(),
+          get_ntp_config().is_archival_enabled(),
+          get_ntp_config().is_read_replica_mode_enabled());
+        co_await _cloud_storage_partition->erase();
+    } else {
+        vlog(
+          clusterlog.info, "Leaving S3 objects behind for partition {}", ntp());
+    }
```
I am not sure if this code should be part of `remove_persistent_state` this method is called when a partition is moved between nodes (it previously removed only local state) Now it will remove data from object store

mmaslankaprv commented:
Would it be possible to add a test case where we validate if data in S3 are available after partition is being moved cross nodes ?

jcsp commented:
- Added `TopicDeleteCloudStoragetest.partition_movement_test`
- Extended `CreateTopicUpgradeTest` to check for data non-deletion on legacy topics, as well as just checking they have remote.delete set the correct value.
- In `CreateTopicsTest`, cover setting remote.delete at creation time.
- Fix setting property on topic creation (by adding to `to_cluster_type`, `from_cluster_type`)
- Fix deleting data incorrectly on partition moves, by adding a `partition_removal_mode` that controller_backend can set appropriately based on whether it's doing a configuration update or an actual deletion.
- Fix sometimes missing the correct value of remote.delete in cluster::partition, by moving the property into `ntp_config` so that it doesn't need special handling and gets passed around/updated along with other config.

mmaslankaprv commented on src/v/cluster/partition_manager.cc:
```diff
@@ -246,6 +247,13 @@ ss::future<> partition_manager::remove(const model::ntp& ntp) {
+      .then([partition, mode] {
+          if (mode == partition_removal_mode::global) {
+              return partition->remove_remote_persistent_state();
+          } else {
```
nit: no need for else

mmaslankaprv commented on src/v/kafka/server/handlers/topics/types.cc:
```diff
@@ -185,6 +185,10 @@ to_cluster_type(const creatable_topic& t) {
       = get_tristate_value<std::chrono::milliseconds>(
         config_entries, topic_property_retention_local_target_ms);
 
+    cfg.properties.remote_delete
```
can this change be squashed with handling for the property in create topics request ?

jcsp commented:
CI failure was https://github.com/redpanda-data/redpanda/issues/6810
